File: 240109-search-sort/2-15-quick-sort.c
#include <stdio.h>
 
 int partition(int items[], int left, int right) {
    int key = items[right];
    int i = left - 1;
        for (int j = left; j < right; j++) {
            if (items[j] <= key) {
                i++;
                int temp = items[i];
                items[i] = items[j];
                items[j] = temp;
            }
        }
        int temp = items[i + 1];
        items[i + 1] = items[right];
        items[right] = temp;
        return (i + 1);
}
 
void quickSort(int arr[], int left, int right) {
    if (left < right) {
        int keyIndex = partition(arr, left, right);
        quickSort(arr, left, keyIndex - 1);// Sort the elements on the left of a key element
        quickSort(arr, keyIndex + 1, right);// Sort the elements on the right of a key element
        }
}
 
void showArray(int arr[], int size) 
{ 
    int i; 
    for (i=0; i < size; i++) 
        printf("%d ", arr[i]); 
    printf("\n"); 
}
 
int main()
{ 
    int libraryNum[] = {124,235,456,123,756,476,285,998,379,108}; 
    int n = sizeof(libraryNum)/sizeof(libraryNum[0]); 
 
    printf("Initial array: \n"); 
    showArray(libraryNum, n); 
    quickSort(libraryNum, 0, n-1); 
    printf("Sorted array: \n"); 
    showArray(libraryNum, n); 
    return 0; 
}

File: 240109-search-sort/2-15-quick-sort.py
def partition(arr,left,right): 
    i = ( left-1 )      # index of smaller element 
    key = arr[right]    # a key element 
 
    for j in range(left, right): 
        if   arr[j] < key:  
            i += 1 
            arr[i],arr[j] = arr[j],arr[i]  
    arr[i + 1],arr[right] = arr[right],arr[i + 1] 
    return ( i+1 )
 
def quick_sort(arr,left,right):
    if left < right:
        key_index = partition(arr,left,right)
        quick_sort(arr,left,key_index - 1)
        quick_sort(arr,key_index + 1, right)        
 
library_num = [124,235,456,123,756,476,285,998,379,108]
print("Initial array:", library_num)
quick_sort(library_num, 0, len(library_num) - 1)
print ("Sorted array:", library_num)